src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs

```diff
@@ -4291,15 +4287,13 @@ .maxstack  2
   IL_0019:  ldc.i4.m1
   IL_001a:  stfld      ""int C.<G>d__0.<>1__state""
   IL_001f:  ldloc.0
-  IL_0020:  ldfld      ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder<object> C.<G>d__0.<>t__builder""
-  IL_0025:  stloc.1
-  IL_0026:  ldloca.s   V_1
-  IL_0028:  ldloca.s   V_0
-  IL_002a:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder<object>.Start<C.<G>d__0>(ref C.<G>d__0)""
-  IL_002f:  ldloc.0
-  IL_0030:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder<object> C.<G>d__0.<>t__builder""
-  IL_0035:  call       ""System.Threading.Tasks.Task<object> System.Runtime.CompilerServices.AsyncTaskMethodBuilder<object>.Task.get""
-  IL_003a:  ret
+  IL_0020:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder<object> C.<G>d__0.<>t__builder""
```

Comment:
This IL seems weird. As far as I can tell `C.<G>d__0` is a value type and it is loaded on the stack by the above `ldloc.0`. So how come `ldflda` gets emitted here, considering that `ldflda` expects and address on the stack?

Reply:
I see, this is compiled with optimizations disabled so `<G>d__0` is really a class.

Reply:
Is the address of the field `<>t__builder` of the local `<G>d__0` which it then invokes a method on (`<>t__builder.Start(ref ...)`) passing in the local as the `ref` param ?
```csharp
private struct <G>d__0 : IAsyncStateMachine
{
    public int <>1__state;
    public AsyncTaskMethodBuilder<object> <>t__builder;
    public Task<object> o;
    private TaskAwaiter<object> <>u__1;
}
```
Or do you mean something else? My il isn't great...

Reply:
Yes, that's the address of `<>t__builder` that's used as `this` for `Start`. Since this is compiled without optimizations the state machine is a class, not a struct, so `ldloc.0` loads a reference to the state machine object and then `ldflda` loads the address of the builder field.

Reply:
There's still a lot of over zeroing that goes on; as there's the zeroing in header `rep stosd` and then the overwrite with `AsyncTaskMethodBuilder<int>.Create()` which is also zeroing. So I did wonder about moving the `.Create()` first; so perhaps the Jit could detect more easily that the zeroing at the start of `G_M56410_IG02` was unnecessary?

```asm
G_M56410_IG01:
       push     rdi
       push     rsi
       sub      rsp, 88
       vzeroupper 
       mov      rsi, rcx
       lea      rdi, [rsp+28H]
       mov      ecx, 12
       xor      rax, rax
       rep stosd 
       mov      rcx, rsi
       mov      rsi, rcx
						;; bbWeight=1    PerfScore 30.00
G_M56410_IG02:
       xor      ecx, ecx
       vxorps   xmm0, xmm0
       vmovdqu  xmmword ptr [rsp+30H], xmm0
       vmovdqu  xmmword ptr [rsp+40H], xmm0
       mov      qword ptr [rsp+50H], rcx
       mov      dword ptr [rsp+2CH], edx
       mov      dword ptr [rsp+28H], -1
       lea      rcx, [rsp+28H]
       call     System.Runtime.CompilerServices.AsyncMethodBuilderCore:Start(byref)
```

Wasn't just if that was just a remote hope though; so didn't reorder.

Reply:
Well, it is what it is. JIT's handling of structs has been poor for the past 20 years...

I don't know why `Create` exists, if anyone expects it to ever return something other that `default`. Otherwise it would be simpler to just not call it.

Reply:
Was an issue raised that any state change done in create doesn't "stick" so someone was trying to call it https://github.com/dotnet/roslyn/issues/41222

C# compiler complains if you don't call it due to non fully initialized struct (otherwise you have to set the MethodBuilder to default which then introduces another set of zeroing). However; don't think it needs to be called from the il?

@stephentoub any thoughts on `.Create()` could it be a non-call like `.SetStatemachine` has become?

Reply:
I think that'd be a breaking change that's likely to break someone, especially if someone's written a builder that's a reference type rather than a value type.

With SetStateMachine, the changes we made are local to our own implementation, and don't actually bleed out into the contract; changing what the compiler generates around calling Create would impact the contract pretty significantly.

Reply:
So maybe moving `.Create` first, so it immediately comes after the zeroing in the preamble, then highlight in [dotnet/runtime](https://github.com/dotnet/runtime) that the Jit isn't eliding the zeroing at the start (from Create); when its already been done and there is no extra work (e.g. setting state to -1) between the two; might be a way to go instead?